crates/bevy_transform/src/systems.rs

```diff
@@ -89,100 +145,423 @@ pub fn propagate_transforms(
+        // number of channel sends by avoiding sending partial batches.
+        queue.send_batches();
+
+        // Spawn workers on the task pool to recursively propagate the hierarchy in parallel.
+        let task_pool = ComputeTaskPool::get_or_init(TaskPool::default);
+        task_pool.scope(|s| {
+            (0..task_pool.thread_num())
+                .for_each(|_| s.spawn(async { propagation_worker(&queue, &nodes) }));
+        });
+    }
+
+    /// A parallel worker that will consume processed parent entities from the queue, and push
+    /// children to the queue once it has propagated their [`GlobalTransform`].
+    #[inline]
+    fn propagation_worker(queue: &WorkQueue, nodes: &NodeQuery) {
+        #[cfg(feature = "std")]
+        let _span = bevy_log::info_span!("transform propagation worker").entered();
+
+        let mut outbox = queue.local_queue.borrow_local_mut();
+        loop {
+            // Try to acquire a lock on the work queue in a tight loop. Profiling shows this is much
+            // more efficient than relying on `.lock()`, which causes gaps to form between tasks.
+            let Ok(rx) = queue.receiver.try_lock() else {
+                core::hint::spin_loop(); // No apparent impact on profiles, but best practice.
+                continue;
+            };
+            // If the queue is empty and no other threads are busy processing work, we can conclude
+            // there is no more work to do, and end the task by exiting the loop.
+            let Some(mut tasks) = rx.try_iter().next() else {
+                if queue.busy_threads.load(Ordering::Relaxed) == 0 {
+                    break; // All work is complete, kill the worker
+                }
+                continue; // No work to do now, but another thread is busy creating more work.
+            };
+            if tasks.is_empty() {
+                continue; // This shouldn't happen, but if it does, we might as well stop early.
+            }
+
+            // If the task queue is extremely short, it's worthwhile to gather a few more tasks to
+            // reduce the amount of thread synchronization needed once this very short task is
+            // complete.
+            while tasks.len() < WorkQueue::CHUNK_SIZE / 2 {
+                let Some(mut extra_task) = rx.try_iter().next() else {
+                    break;
+                };
+                tasks.append(&mut extra_task);
+            }
+
+            // At this point, we know there is work to do, so we increment the busy thread counter,
+            // and drop the mutex guard *after* we have incremented the counter. This ensures that
+            // if another thread is able to acquire a lock, the busy thread counter will already be
+            // incremented.
+            queue.busy_threads.fetch_add(1, Ordering::Relaxed);
+            drop(rx); // Important: drop after atomic and before work starts.
```

Comment:
Can we move the work queue logic to `bevy_tasks` or some place? Normally, I'm not big on generalizing prematurely, but from a code organization perspective it feels better to have `bevy_transform` be for higher level logic and `bevy_tasks` (or whatever) be for low-level parallelism nuts and bolts.

Reply:
I tried to do this, but it ended up being a pretty major time sink, and I couldn't quite figure out how to express the lifetimes with generic queries. I've mentioned this exists to Alice to keep it in mind during future work. `nth` is working on some scheduler improvements, and is also aware of this work. I think that is probably a decent place to leave this for now, I might try extracting this logic again if I can make some time, but it's possible the work `nth` is doing will be usable here similar to rayon.